KAFKA-16123: KStreamKStreamJoinProcessor does not drop late records.

florin-akermann opened the pull request:
KStreamKStreamJoinProcessor forwards null records for left/outer joins unconditionally of the join window.

florin-akermann committed:
KAFKA-16123: Fix KStreamKStreamJoinProcessor forwards null records for left/outer joins unconditionally of the join window.

florin-akermann committed:
KAFKA-16123: Generalize active window check for all types of records

florin-akermann commented:
This might actually be a general issue and not just for null-key records?
In other words, the problem already existed prior to KIP-962 for non null-key records.

See https://github.com/apache/kafka/pull/15314/files

In other words, the fix needs to be generalized (not just for null-key records)

florin-akermann commented:
Testing strategy:
Most test cases can be left intact by adding a large enough grace period. In general, the grace period added should be as small as possible.

Additional test cases to test the behavior for grace expiry edge cases have been added.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -191,6 +191,10 @@ public void process(final Record<K, V1> record) {
         }
 
+        private boolean isActiveWindow(final long timeFrom, final long timeTo) {
+            return sharedTimeTracker.streamTime >= timeFrom && timeTo + joinGraceMs >= sharedTimeTracker.streamTime;
```
Do we need to check `sharedTimeTracker.streamTime >= timeFrom` ?

If I am not mistaken, `timeTo` is always larger that `timeFrom`, and `graceMs` is always non-negative. Thus, if `timeTo + joinGraceMs >= sharedTimeTracker.streamTime;` it implies that `sharedTimeTracker.streamTime >= timeFrom` is `true`? And if `timeTo + joinGraceMs >= sharedTimeTracker.streamTime;` is false, the whole condition is false independent of `sharedTimeTracker.streamTime >= timeFrom`?

florin-akermann replied:
yes, indeed, it is redundant to check for `sharedTimeTracker.streamTime >= timeFrom`

adjusted

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java:
```diff
@@ -128,7 +128,7 @@ public void process(final Record<K, V1> record) {
             final long timeTo = Math.max(0L, inputRecordTimestamp + joinAfterMs);
             sharedTimeTracker.advanceStreamTime(inputRecordTimestamp);
-
+            if (!isActiveWindow(timeFrom, timeTo)) return;
```
nit: code style -- we never use blocks without `{ }` -- this should be
```
    if (!isActiveWindow(timeFrom, timeTo)) {
        return;
    }
```

Also, I think we should record this and call `StreamStreamJoinUtil.skipRecord(record, LOG, droppedRecordsSensor, context())` similar to key and/or valud being `null`.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/RelaxedNullKeyRequirementJoinTest.java:
```diff
@@ -71,11 +71,21 @@ void afterEach() {
+        );
     }
 
     @Test
```
Not related to this PR, but the new stream-table join with versioned state stores also has a grace-period -- can we double check the KIP if we decided to drop stream-records for left-join when they arrive after grace-period and maybe extend this test below (ie `testRelaxedLeftStreamTableJoin`) accordingly (in a follow up PR)?

florin-akermann replied:
Yes, I will check.

Similarily, I realized `KStreamKStreamSelfJoin` shoud probably also drop 'too late' records? I guess this would also be  a separate Jira ticket & PR?

mjsax replied:
Yes it should, and yes, let's keep it separate -- if you want to extend the jira for self-joins or create a new one does not matter too much. We can also do more than one PR for a single Jira -- but this PR is already complex enough, so let's not add the self-join stuff, but only tackle it after we merged this one.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:
```diff
@@ -816,10 +816,12 @@ public void testWindowing() {
         stream2 = builder.stream(topic2, consumed);
 
+        final Duration timeDifference = ofMillis(100L);
+        final Duration grace = ofMillis(104);
```
I think we need a larger grace period -- this test aims to test window bounds, and no input data should be dropped as late. -- We should set it to `150`? -- Can we also add a check on the dropped record sensor that it does report zero? (Also cf my other commend below)

florin-akermann replied:
Good idea with the sensor.

Just for me to understand, is 150 an arbitrarily chosen value or how did you come up with it?
Wouldn't it make sense to set it to at least (max_timestamp in test case - min_timestamp in test case) e.g. 1104 - 899?

florin-akermann replied:
On second thought, note my coment below on 'Why + 1 '.

mjsax replied:
Yes, `150` was random -- just a "round" number that is big enough. -- Don't think we need to set a "strict" bound by computing the minimum grace period we need (IMHO, it would just be distracting if one read the test to figure out where `1104 - 899` comes from...

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:
```diff
@@ -1363,6 +1365,16 @@ public void testWindowing() {
                 new KeyValueTimestamp<>(3, "L3+l3", 2003L)
             );
+
+            //push two items with timestamp at grace edge; this should produce one join item, M0 is 'too late'
```
We should add the "drop record sensor check == 0" before we intentionally push "late" data that should be dropped, and we should also test the sensor at the very end that is recorded the dropped records.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:
```diff
@@ -1363,6 +1365,16 @@ public void testWindowing() {
+            final long currentStreamTime = 2104;
+            final long lowerBound = currentStreamTime - timeDifference.toMillis() - grace.toMillis();
+            inputTopic1.pipeInput(0, "M0", lowerBound - 1);
+            inputTopic1.pipeInput(1, "M1", lowerBound + 1);
```
Why `+ 1` -- `lowerBound` by itself should already be inclusive and produce a join result?

florin-akermann replied:
I can see now that the naming is misleading.

The lowerbound is with regards to the grace period, 1900
However the lowerbound of the winow `1:l1` is 1901

So the +1 was there to make sure it is still within the window.

In general I start to wonder whether it wouldn't make more sense to test these two concerns (grace & windowing) separatley. E.g. with grace 150, `M0` is just a test case to assert that late records get dropped and `M1` is just another window bound test. With grace 104 we get the 'grace bound' and the 'l0 lower window bound' to overlap but it might be confusing. In other words, as you said 'this test aims to test window bounds'. So maybe I should move grace period tests into a separate test class?

mjsax replied:
> The lowerbound is with regards to the grace period, 1900
> However the lowerbound of the winow 1:l1 is 1901

> So the +1 was there to make sure it is still within the window.

For this case, should `M0` not have the `-1` to get a strict bound test? Comparing the other test code you updated, you also use `lowerBound -1` to a "late" record and `lowerBound` for still within grace period record. -- Thus, should `currentStreamTime = 2105` to make it all work as expected? -- There is some disconnect between this test setup and other test setup.

mjsax replied:
> E.g. with grace 150, M0 is just a test case to assert that late records get dropped and M1 is just another window bound test. With grace 104 we get the 'grace bound' and the 'l0 lower window bound' to overlap but it might be confusing. In other words, as you said 'this test aims to test window bounds'. So maybe I should move grace period tests into a separate test class?

Well, yes and no. I general, I agree that we should "separate" window bound and grace-period test, but I believe we actually do this already -- if we use grace 150, we don't worry about "late" data that might be dropped in the first phase of the test. And we bolt-on a last test case for grace (I don't think we would gain much to separate this into a single test method). -- The confusing part might be, that we cannot clearly "only test grace period" w/o also (implicitly) test window bound... The grace test shows us, that a "late" record is dropped, and that a "still within grace bounds record is not dropped and this "still in grace bounds record" naturally falls at the window-start time of the window it belong into (it's the oldest window that is still open, and it sell accept records with ts = window-start)

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:
```diff
@@ -1382,7 +1394,7 @@ public void testAsymmetricWindowingAfter() {
             stream2,
             MockValueJoiner.TOSTRING_JOINER,
-            JoinWindows.ofTimeDifferenceWithNoGrace(ofMillis(0)).after(ofMillis(100)),
+            JoinWindows.ofTimeDifferenceAndGrace(ofMillis(0), ofMillis(4)).after(ofMillis(100)),
```
As above: this test is an inner-join and setup to never drop record due to grace -- we should set a larger value and also check "dropped record sensor".

It would also make sense to add the same extension to send additional "late" data at the end of this test?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:
```diff
@@ -1651,7 +1663,7 @@ public void testAsymmetricWindowingBefore() {
             stream2,
             MockValueJoiner.TOSTRING_JOINER,
-            JoinWindows.ofTimeDifferenceWithNoGrace(ofMillis(0)).before(ofMillis(100)),
+            JoinWindows.ofTimeDifferenceAndGrace(ofMillis(0), ofMillis(4)).before(ofMillis(100)),
```
as above

mjsax reviewed (commented):
Thanks for the fix! Overall LGTM. Couple of comments.

florin-akermann committed:
KAFKA-16123
- log skip
- notify dropped record sensor

florin-akermann committed:
KAFKA-16123: Use the largest of graceBound and windowBound for test case 'M1' as timestamp

florin-akermann commented:
> Thanks for the fix! Overall LGTM. Couple of comments.

@mjsax Thank you for all the good points. I agree with all of them. However, first I would like to align on whether we really want to extend the window bound tests to assert the correct grace period behavior. See my reply on 'Why + 1'.

Personally, I would say I just write separate TestClasses/Cases to assert the grace period behavior (including sensor checks).

Plus, as you suggested, the window bound tests will just be updated with large grace periods (E.g. Long.MAX - some constant).

mjsax commented:
Replies in-line -- in the end I don't have a strong opinion -- if you think it's better to use grace=MAX (instead of just "large enough 150) and add a new tests method just for grace, I am also happy with it.

florin-akermann committed:
KAFKA-16123: Assert records arriving post window close are dropped

florin-akermann committed:
KAFKA-16123: rm line

florin-akermann committed:
KAFKA-16123: Alternative test setup for late record drop behavior. If only assert the dropped record sensor state and not the join result (which is asserted by other tests already) then we could use an identical test for the different stream operators (inner, left, outer)

florin-akermann committed:
KAFKA-16123: adjust skip record log statement

florin-akermann committed:
KAFKA-16123: outerJoin

mjsax commented:
@florin-akermann -- I finally merged https://github.com/apache/kafka/pull/14426 -- can you rebase this PR and fixup tests so we can move forward with this PR?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamStreamJoinUtil.java:
```diff
@@ -25,36 +25,46 @@
+        if (context.recordMetadata().isPresent()) {
+            final RecordMetadata recordMetadata = context.recordMetadata().get();
+            logger.warn(
+                "Skipping record. reason=[{}] topic=[{}] partition=[{}] offset=[{}]",
```
Nit: Just comparing to `AbstractKStreamTimeWindowAggregateProcessor#logSkippedRecordForExpiredWindow` is seems we could addd more information -- should we try to merge both "skip reason" as your PR proposes, if have two different output, one key null-case, and a different one for "expired" case similar to windowed-aggregation?

mjsax replied:
```suggestion
                "Skipping record. Reason=[{}] topic=[{}] partition=[{}] offset=[{}]",
```

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/RelaxedNullKeyRequirementJoinTest.java:
```diff
@@ -71,11 +71,21 @@ void afterEach() {
+            Arrays.asList(
+                new KeyValue<>(null, "leftValue1|null"),
+                new KeyValue<>(null, "leftValue2|null"),
+                new KeyValue<>(null, "lateArrivalWithinGrace|null")
```
Why is this record already in the output? We should not drop it, but it seems we cannot emit it right away either, because we need to wait until the window closes, so would need to pipe one more record with ts=91 to flush out this result?

florin-akermann replied:
Yes, good point. As is, the KStreamKStreamJoin forwards null-key records directly and not only after the window closes.

florin-akermann replied:
I wonder how to best solve this.
Multiple null-key records would colide in the 'outerJoinStore' as they potentially all could have the same key of  type 'TimestampedKeyAndJoinSide<K>'.

E.g. for  a left stream

| Key  | value | ts |
|------|-------|----|
| null | a     | 1  |
| null | b     | 1  |
| null | c     | 1  |

We probably would only get to see 'c' even though we would like to see 'a','b' and 'c'?

From the top of my head I see two options to handle this.
- Maintain an additional store just for null-key records where such records wouldn't collide.
- Adjust the 'outerJoinStore' key type `TimestampedKeyAndJoinSide<K>'`( E.g. by adding offset and partition as optional fields. This way null-key records could be distinguished and for keyed records the old behavior can be kept)

Personally I prefer the latter.

florin-akermann replied:
I just realized that partition and offset are not available in this context...
So, with regards to the second bullet point, probably the only distinguishing feature would be the value itself?

I start to wonder, is it really worth it to add this complexit? We know all three events should eventually be sent downstream anyway. This way we could avoid using additional storage & extending `TimestampedKeyAndJoinSideSerde` with another generic field.

mjsax replied:
Oh. This is `null`-key case. My bad. For `null`-key we know it won't join in the future, so no reason to artificially delay the output. Thanks for pointing it out.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/RelaxedNullKeyRequirementJoinTest.java:
```diff
@@ -91,13 +101,22 @@ void testRelaxedLeftStreamTableJoin() {
+                new KeyValue<>(null, "null|rightValue"),
+                new KeyValue<>(null, "leftValue|null"),
+                new KeyValue<>(null, "null|lateArrivalWithinGrace"),
+                new KeyValue<>(null, "lateArrivalWithinGrace|null")
```
As above for last two records -- we should not yet seem them, without advancing time to `91`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoinTest.java:
```diff
@@ -1901,6 +1903,66 @@ public void testAsymmetricWindowingBefore() {
+            right.pipeInput(-1, "bumpTime", 40);
+            assertRecordDropCount(0.0, processor);
+
+            right.pipeInput(0, "closesAt39", 24);
```
This should be `closeAt40`? The window is from `[24-10; 24+10] = [14;34]` (note, both bounds inclusive), so with zero grace, we would close at 35, but with 5 grace, we close at 40.

florin-akermann replied:
Thanks, ok, I adjuted the 'hint' in the value accordingly.
I don't think we have off-by-one issue here: `[14;34 + 5]` so the record is considered 'too late' at t=40?
In other words for this test case it was purely a misleading 'hint'?

On a different note, I deleted the test case in `KStreamKStreamJoinTest` and refer to `KStreamKStreamWindowCloseTest`.

mjsax reviewed (commented):
Thanks for the update. It seems there is an off-by-one bug when we actually close a window?

About existing tests: yes, just adding a grace-period is fine with me, however, it should impact when a left/right join result is emitted, right? So I would have expected that some test need an update with either advancing time pro-actively, or by expected certain result later in the test, because windows are closes later?

florin-akermann committed:
KAFKA-16123: KStreamKStreamWindowCloseTest adjust window close 'hint in test value

florin-akermann committed:
KAFKA-16123: remove 'recordsArrivingPostWindowCloseShouldBeDropped' from old location

florin-akermann commented:
>  So I would have expected that some test need an update with either advancing time pro-actively, or by expected certain result later in the test, because windows are closes later?

Indeed good point.

I think the problem lies with 
https://github.com/florin-akermann/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java#L247
`outerJoinLeftBreak && outerJoinRightBreak` is always false. Hence, this loop never exits early?
I think this behavior has been introduced as part of https://github.com/apache/kafka/pull/14426?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamWindowCloseTest.java:
```diff
@@ -0,0 +1,150 @@
+
+            right.pipeInput(0, "closedAt40", 24);
+            assertRecordDropCount(1.0, processor);
+
```
If find this test a little bit hard to follow...

We start to open a window `[5;15]` with the first input record. This window would close at `31` so why do we dump to `40`, and test dropping with `ts=24` -- both are totally unrelated to the left input record's window.

Also, it would be good to test that we can join out-of-order record successfully as long as the window is open, and that we don't drop re-mature before we hit the close time, thus I would suggest something like this:
 ```
 // prepare
 - left input at 15 -> open window
 // positive test
 - bump time to 30 (different key) -> window still open
 - right input at 5 -> joins (no need to test with 4, because 4 won't join because of window size anyway...)
 - right input at 25 -> joins 
 // negative test
 - bump time to 31 (different key) -> window closes
 - right input at 5 -> dropped (recorded with metric)
 - right input at 25 -> does not join any longer; window closed (for this case we don't drop and don't record metric)
 // test sharp lower drop bound
  - right input at 6 -> does not join any longer; window closed (for this case we don't drop though and don't record metric)
// cont. with additional sanity check:
 - left input at 16 -> joins with both right input at 6 and right input at 25 -- to verify both records did not get dropped
 ```

florin-akermann replied:
Ok interesting,  given your test cases, I think misunderstood when we get to drop records.
To check `sharedTimeTracker.streamTime > timeTo + joinGraceMs` ([here](https://github.com/apache/kafka/blob/9e529373df2d157304c03b713b999c2bb1300613/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java#L132)) seems to be too simplistic.
Because this would already drop the first `right input at 5` beccause (5 + 10 + 5 < 30).
So now I am thinking it needs to include both records in the evaluation.
Something like:
For a matching left and right record only drop the current record if 
`max(thisRecord.ts , otherRecord.ts) + joinAfterMs + grace < sharedTimeTracker.streamTime`.
And that logic would be applied [here](https://github.com/apache/kafka/blob/9e529373df2d157304c03b713b999c2bb1300613/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java#L154)

Long story short, I will work on this but wont be able to do so before end of next week... I hope that is ok.

mjsax replied:
Thanks. No rush! We still have time util 3.8 comes around :)

For joins it's a little bit more complicated than for aggregations, but in the end, we need to accept a record as along as a window is open, and window-close time is defined as window-end+grace+1 (for join windows). And for joins, the definition of what a window is, is a little bit more complex (especially when taking "asymmetric" windows into account).

I guess bottom line is, that every input record, does conceptually open a join window over the other input stream, and we need to keep this window open an accept records from the other side until we reach window close time. At the same time, if an out-of-order record comes in and we put it into an existing join window, it might be allowed to open it's own join window any longer, because it's own join window is already too old (this is the weird case that cannot happen for aggregations...)

I did not think about the right/best way how to compute this correctly, but I am sure you'll figure it out :)

florin-akermann committed:
KAFKA-16123: Update StreamStreamJoinUtil

mjsax commented:
Just merged https://github.com/apache/kafka/pull/15510 -- can we move forward with this PR (maybe by rebasing it to see if any tests break?)?

florin-akermann commented:
Hi @mjsax , thanks for the flag.
Yes I'll push the necessary changes by the end of the week (Sunday).
I hope that's ok.

florin-akermann committed:
KAFKA-16123: KStreamKStreamJoinTest, distinguish cases: late record (drop) vs closed window

florin-akermann committed:
KAFKA-16123: move test cases to KStreamKStreamWindowCloseTest

florin-akermann committed:
KAFKA-16123: Add asymmetric test case scenario.

florin-akermann committed:
KAFKA-16123: Drop null key record when it's own window is close.

florin-akermann committed:
KAFKA-16123: Adjust record skip reason.

florin-akermann commented:
Hi @mjsax 
It's now rebased.
Plus, I addressed your reply below.

> If find this test a little bit hard to follow...
> 
> We start to open a window `[5;15]` with the first input record. This window would close at `31` so why do we dump to `40`, and test dropping with `ts=24` -- both are totally unrelated to the left input record's window.
> 
> Also, it would be good to test that we can join out-of-order record successfully as long as the window is open, and that we don't drop re-mature before we hit the close time, thus I would suggest something like this:
> 
> ```
> // prepare
> - left input at 15 -> open window
> // positive test
> - bump time to 30 (different key) -> window still open
> - right input at 5 -> joins (no need to test with 4, because 4 won't join because of window size anyway...)
> - right input at 25 -> joins 
> // negative test
> - bump time to 31 (different key) -> window closes
> - right input at 5 -> dropped (recorded with metric)
> - right input at 25 -> does not join any longer; window closed (for this case we don't drop and don't record metric)
> // test sharp lower drop bound
>  - right input at 6 -> does not join any longer; window closed (for this case we don't drop though and don't record metric)
> // cont. with additional sanity check:
> - left input at 16 -> joins with both right input at 6 and right input at 25 -- to verify both records did not get dropped
> ```

I would argue that the second 'right input at 25' still should join / emit a record because the other window is still active and 'ts=15' within its bounds. Other than that the expected behavior of the test cases in KStreamKStreamWindowCloseTest is in line with what you suggested.

florin-akermann committed:
KAFKA-16123: adjust join skip predicate
